### DISCUSSION TITLE
Why `command echo -e` is not working correctly?

### DISCUSSION THREAD

**Discussion by julyfun** [fish-shell/fish-shell#10493] [Q&A] [answered]
os: macos 14.5
fish shell version: 3.7.1

```
# in bash:
bash-5.2$ command echo -e 123
123
# in fish:
~ λ command echo -e 123
-e 123
~ λ command echo -n 123
123
~ λ echo -e 123
123
```

All but `command echo -e 123` in fish shell works correctly, the `-e` should be a parameter but it is in the output.

**Comment by faho** [comment, marked as answer]
Bash's `command` builtin will only ignore functions, not builtins. See https://www.gnu.org/software/bash/manual/html_node/Bash-Builtins.html. You can also run `command -V echo` (yes, uppercase "V") in bash to see that it prints "echo is a shell builtin". So `command echo` in bash still runs bash's version of echo.

In fish, `command` will only run external commands, so in this case it will run something like `/usr/bin/echo`.

If `command echo -e 123` in fish prints the `-e`, then that is because that's what the echo binary on your system does. `echo` is annoyingly unportable because implementations work differently in these things. In this case that `echo` simply does not have the `-e` option, and so it prints it (posix also forbids it from supporting a `--` option separator, so it is *impossible* to print arbitrary strings with `echo` portably in posix-conforming systems).

(tbh I've always thought that bash's `command` was extremely confused about what it actually is *for* - if you want to run the builtin echo, why would you not use `builtin echo`?)
